### PR TITLE
Uniformize Kubernetes api versions

### LINKED ISSUE #9165: Use correct api versions for Kubernetes
Our usage of apiVersion is very inconsistent, For some of the generated Kubernetes files we are still using `extensions/v1beta1`. For services we use `apiVersion: v1` and for some deployments `apps/v1` I think we need to review and update them. Especially the SQL database maifests. See below for example

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: invoice-mysql
  namespace: xl-demo
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: invoice-mysql
    spec:
      volumes:
      - name: data
        emptyDir: {}
      containers:
      - name: mysql
        image: mysql:5.7.20
        env:
        - name: MYSQL_USER
          value: root
        - name: MYSQL_ALLOW_EMPTY_PASSWORD
          value: 'yes'
        - name: MYSQL_DATABASE
          value: invoice
        args:
        - --lower_case_table_names=1
        - --skip-ssl
        - --character_set_server=utf8mb4
        - --explicit_defaults_for_timestamp
        ports:
        - containerPort: 3306
        volumeMounts:
        - name: data
          mountPath: /var/lib/mysql/
---
apiVersion: v1
kind: Service
metadata:
  name: invoice-mysql
  namespace: xl-demo
spec:
  selector:
    app: invoice-mysql
  ports:
  - port: 3306

```
@PierreBesson @srinivasa-vasu could you guys have a glance

### PR DESCRIPTION
**Pull request by PierreBesson**
Fixes #9165

### CODE REVIEW

**Comment by murdos** [comment]
Yeah, less duplicated code! 👍

**Comment by DanielFran** [comment]
@PierreBesson Is there conflits with #8987?

**Comment by pascalgrimaud** [comment]
https://github.com/jhipster/generator-jhipster/pull/8987 has already conflicts

**Comment by PierreBesson** [comment]
Can someone review this please ?

**Comment by pascalgrimaud** [comment]
I'm not a Kubernetes expert but I can try and deploy to GCP :)
Let me test it

File: generators/kubernetes-base.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    });
+}
+
+function loadConfig() {
+    loadFromYoRc.call(this);
+    this.kubernetesNamespace = this.config.get('kubernetesNamespace');
+    this.kubernetesServiceType = this.config.get('kubernetesServiceType');
+    this.ingressDomain = this.config.get('ingressDomain');
+    this.istio = this.config.get('istio');
+    this.istioRoute = this.config.get('istioRoute');
```
Comment (DanielFran) [inline generators/kubernetes-base.js]: Istioroute has been removed

File: generators/kubernetes-base.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        serviceDiscoveryType: this.serviceDiscoveryType,
+        jwtSecretKey: this.jwtSecretKey,
+        dockerRepositoryName: this.dockerRepositoryName,
+        dockerPushCommand: this.dockerPushCommand,
+        kubernetesNamespace: this.kubernetesNamespace,
+        kubernetesServiceType: this.kubernetesServiceType,
+        ingressDomain: this.ingressDomain,
+        monitoring: this.monitoring,
+        istio: this.istio,
+        istioRoute: this.istioRoute
```
Comment (DanielFran) [inline generators/kubernetes-base.js]: Istioroute has been removed

File: generators/kubernetes-helm/files.js [hunk 3]
```diff
@@ -24,47 +24,53 @@ module.exports = {
...
-                    this.template(`${k8s}/db/elasticsearch.yml.ejs`, `${appName}/templates/${appName}-elasticsearch.yaml`);
+                    this.template(`${kubernetesSubgenPath}/db/elasticsearch.yml.ejs`, `${appName}/templates/${appName}-elasticsearch.yaml`);
                 }
                 if (
                     (this.app.applicationType === 'gateway' || this.app.applicationType === 'monolith') &&
                     this.kubernetesServiceType === 'Ingress'
                 ) {
                     if (this.istio !== 'no') {
-                        this.template(`${k8s}/gateway.yml.ejs`, `${appName}/templates/${appName}-gateway.yaml`);
+                        this.template(`${kubernetesSubgenPath}/gateway.yml.ejs`, `${appName}/templates/${appName}-gateway.yaml`);
```
Comment (DanielFran) [inline generators/kubernetes-helm/files.js]: File has been mover to istio folder

File: generators/kubernetes-helm/files.js [hunk 4]
```diff
@@ -24,47 +24,53 @@ module.exports = {
...
 
                 if (this.app.searchEngine === 'elasticsearch') {
-                    this.template(`${k8s}/db/elasticsearch.yml.ejs`, `${appName}/templates/${appName}-elasticsearch.yaml`);
+                    this.template(`${kubernetesSubgenPath}/db/elasticsearch.yml.ejs`, `${appName}/templates/${appName}-elasticsearch.yaml`);
                 }
                 if (
                     (this.app.applicationType === 'gateway' || this.app.applicationType === 'monolith') &&
                     this.kubernetesServiceType === 'Ingress'
                 ) {
                     if (this.istio !== 'no') {
```
Comment (DanielFran) [inline generators/kubernetes-helm/files.js]: Istio property os now a boolean

File: generators/kubernetes-helm/files.js (hunk 3 above)
Comment (PierreBesson) [reply generators/kubernetes-helm/files.js]: OK

File: generators/kubernetes-base.js (hunk 1 above)
Comment (PierreBesson) [reply generators/kubernetes-base.js]: OK, I removed it.

File: generators/kubernetes-helm/files.js (hunk 4 above)
Comment (PierreBesson) [reply generators/kubernetes-helm/files.js]: OK, rebased

File: generators/kubernetes-base.js (hunk 2 above)
Comment (PierreBesson) [reply generators/kubernetes-base.js]: OK

**Comment by DanielFran** [comment]
@PierreBesson did you added yarn.lock filé by mistake?

**Comment by PierreBesson** [comment]
Yes I did.

**Comment by PierreBesson** [comment]
@PierreBesson I have remove the package-lock.

**Comment by jdubois** [comment]
I'm adding a bug bounty here, as I want this for JHipster 6

**Comment by DanielFran** [comment]
@PierreBesson Sorry, did not have time this week to test.

**Comment by pascalgrimaud** [comment]
@PierreBesson : I'm testing your PR, and I got this issue with MongoDB. I used the same `workshop.jdl` file :

```
➜ ./kubectl-apply.sh   
configmap/application-config created
secret/registry-secret created
service/jhipster-registry created
statefulset.apps/jhipster-registry created
deployment.apps/accountancy created
configmap/accountancy-mongodb-config created
configmap/accountancy-mongodb-init created
service/accountancy created
error: error validating "accountancy/accountancy-mongodb.yml": error validating data: ValidationError(StatefulSet.spec): missing required field "selector" in io.k8s.api.apps.v1.StatefulSetSpec; if you choose to ignore these errors, turn validation off with --validate=false
deployment.apps/crm created
deployment.apps/crm-elasticsearch created
service/crm-elasticsearch created
secret/crm-postgresql created
deployment.apps/crm-postgresql created
service/crm-postgresql created
service/crm created
deployment.apps/store created
deployment.apps/store-mysql created
service/store-mysql created
service/store created
```

**Comment by pascalgrimaud** [comment]
So now, I got Elasticsearch error, with the famoux `vm.max_map_count`. Here the logs:

```
[2019-03-30T19:26:38,523][INFO ][o.e.b.BootstrapChecks    ] [d36q6Wa] bound or publishing to a non-loopback address, enforcing bootstrap checks
ERROR: [1] bootstrap checks failed
[1]: max virtual memory areas vm.max_map_count [65530] is too low, increase to at least [262144]
[2019-03-30T19:26:38,537][INFO ][o.e.n.Node               ] [d36q6Wa] stopping ...
[2019-03-30T19:26:38,554][INFO ][o.e.n.Node               ] [d36q6Wa] stopped
[2019-03-30T19:26:38,554][INFO ][o.e.n.Node               ] [d36q6Wa] closing ...
[2019-03-30T19:26:38,578][INFO ][o.e.n.Node               ] [d36q6Wa] closed
[2019-03-30T19:26:38,581][INFO ][o.e.x.m.j.p.NativeController] Native controller process has stopped - no new native processes can be started
```

**Comment by pascalgrimaud** [comment]
I applied this change, but the `crm` still failed to start:

```diff
diff --git a/crm/crm-elasticsearch.yml b/crm/crm-elasticsearch.yml
index 08cdecd..e473ff6 100644
--- a/crm/crm-elasticsearch.yml
+++ b/crm/crm-elasticsearch.yml
@@ -16,6 +16,15 @@ spec:
       volumes:
         - name: data
           emptyDir: {}
+      initContainers:
+      - name: init-sysctl
+        image: busybox
+        command:
+        - sysctl
+        - -w
+        - vm.max_map_count=262144
+        securityContext:
+          privileged: true
       containers:
         - name: elasticsearch
           image: docker.elastic.co/elasticsearch/elasticsearch:6.4.3
```

**Comment by pascalgrimaud** [comment]
@PierreBesson : we can review it together this thursday, during our JHipster day

**Comment by PierreBesson** [comment]
@pascalgrimmaud Yes I have encountered the same problem (memory lacking for Elasticsearch) and the strange error for mongodb) but I think those issues where there before my changes.

Basically both Elasticsearch and MongoDb are stateful applications not very suited to run in Kubernetes in the first place. We should migrate to use relevant "operators" to manage them for us. Moreover even with operators, it will be very hard to properly operate them in Kubernetes for a real prod environment.

Let's have a look together next thursday.

**Comment by pascalgrimaud** [comment]
I think I fixed the mongodb issue, and the issue with Elasticsearch with the git diff.
But still have 2 issues:
- crm can't start, I don't understand why
- store starts then be killed, starts again and killed again... so strange

**Comment by pascalgrimaud** [comment]
Sorry, I didn't mention I git commit to your branch. Hope you're ok :)

**Comment by PierreBesson** [comment]
It's totally OK to commit to my branch -> less work for me !
About your issues. What is the size of your cluster. It might be too small for the workload you put on it.

**Comment by pascalgrimaud** [comment]
I created with:
- cluster size: 1
- 4 vCPUs
- 26 Go for total memory

**Comment by PierreBesson** [comment]
Might be not enough CPU. You can try decreasing the cpu reservation in deployments.

**Comment by pascalgrimaud** [comment]
Ok, let me try with a bigger cluster

**Comment by pascalgrimaud** [comment]
It doesn't change anything... Let's see this thursday!

**Comment by pascalgrimaud** [comment]
Just pushed the fix for Elasticsearch memory - cc @Hawkurane

**Comment by pascalgrimaud** [comment]
Found the problem with Elasticsearch.
Here the docs: https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-nosql.html#boot-features-connecting-to-elasticsearch-rest

> Elasticsearch ships two different REST clients that you can use to query a cluster: the "Low Level" client and the "High Level" client.

> If you have the org.elasticsearch.client:elasticsearch-rest-client dependency on the classpath, Spring Boot will auto-configure and register a RestClient bean that by default targets localhost:9200. You can further tune how RestClient is configured, as shown in the following example:

```yaml
spring.elasticsearch.rest.uris=https://search.example.com:9200
spring.elasticsearch.rest.username=user
spring.elasticsearch.rest.password=secret
```

For our config, we didn't do anything about `spring.elasticsearch.rest.uris`, that's why it failed!

**Comment by pascalgrimaud** [comment]
The build failed because of upgrade sub generator. Nothing related to this PR.
That's why I'm merging this.

Congrats for your hard work @PierreBesson 
I was so happy to work on this, with you today !

**Comment by PierreBesson** [comment]
I claimed the bounty : https://opencollective.com/generator-jhipster/expenses/7561

**Pull request merged**